great_expectations/data_context/data_context/abstract_data_context.py

```diff
@@ -602,29 +605,49 @@ def get_datasource(
         self._cached_datasources[datasource_name] = datasource
         return datasource
 
+    def _serialize_substitute_and_sanitize_datasource_config(
+        self, serializer: AbstractConfigSerializer, datasource_config: DatasourceConfig
+    ) -> dict:
+        """Serialize, then make substitutions and sanitize config (mask passwords), return as dict.
+
+        Args:
+            serializer: Serializer to use when converting config to dict for substitutions.
+            datasource_config: Datasource config to process.
+
+        Returns:
+            Dict of config with substitutions and sanitizations applied.
+        """
+        substitutions: dict = self._determine_substitutions()
+        datasource_dict: dict = serializer.serialize(datasource_config)
+
+        substituted_config: dict = cast(
+            dict,
+            substitute_all_config_variables(
+                datasource_dict, substitutions, self.DOLLAR_SIGN_ESCAPE_STRING
+            ),
+        )
+        masked_config: dict = PasswordMasker.sanitize_config(substituted_config)
+        return masked_config
+
     def list_datasources(self) -> List[dict]:
         """List currently-configured datasources on this context. Masks passwords.
 
         Returns:
             List(dict): each dictionary includes "name", "class_name", and "module_name" keys
         """
         datasources: List[dict] = []
-        substitutions: dict = self._determine_substitutions()
 
         datasource_name: str
         datasource_config: DatasourceConfig
+        serializer = ListDatasourceSerializer(schema=datasourceConfigSchema)
```

Comment:
Just brainstorming names since this current one is very strongly coupled with the `list_datasource` method - perhaps something about user friendly, human readable, etc? I'm good with this though.

Reply:
How about `NamedDatasourceSerializer`?